src/adapters/thinkerbell/mod.rs

```diff
@@ -0,0 +1,404 @@
+static ADAPTER_VERSION: [u32;4] = [0, 0, 0, 0];
+
+/// ThinkerbellAdapter hooks up the rules engine (if this, then that) as an adapter.
+///
+/// Each "rule", or "script", is a JSON-serialized structure according to Thinkerbell conventions.
+///
+/// This adapter exposes a root service, with one AddThinkerbellRule setter (to add a new rule).
+/// Each rule that has been added is exposed as its own service, with the following getters/setters:
+/// - Set Enabled (setter) -- toggles whether or not the script is enabled
+/// - Get Enabled (getter) -- returns whether or not the script is enabled
+/// - Remove (setter) -- removes the script
+///
+/// This adapter performs most actions by delegating channel messages to its main thread.
+#[derive(Clone)]
+pub struct ThinkerbellAdapter<T> where T: API + Clone + Send + 'static {
+    /// The sending end of the channel for sending messages to ThinkerbellAdapter's main loop.
+    tx: RawSender<ThinkAction>,
+    /// A reference to the AdapterManager.
+    adapter_manager: T,
+    /// The ID of this adapter (permanently fixed)
+    adapter_id: Id<AdapterId>,
+    /// The ID of the root service's "Add Rule" setter.
+    setter_add_rule_id: Id<Setter>,
+}
+
+/// Thinkerbell requires an execution environment following this API.
+#[derive(Clone)]
+struct ThinkerbellExecutionEnv<T> {
+    adapter_manager: T,
+    // FIXME: Timer's not clonable, so we should only use one, right? Does this have to be mutexed?
+    timer: Arc<Mutex<timer::Timer>>
+}
+
+impl<T> ExecutableDevEnv for ThinkerbellExecutionEnv<T>
+    where T: API + Clone + Send + 'static {
+    // Don't bother stopping watches.
+    type WatchGuard = T::WatchGuard;
+    type API = T;
+
+    fn api(&self) -> Self::API {
+        self.adapter_manager.clone()
+    }
+
+    type TimerGuard = timer::Guard;
+    fn start_timer(&self, duration: Duration, sender: Box<ExtSender<()>>) -> Self::TimerGuard {
+        let timer = self.timer.lock().unwrap(); // FIXME: There's no way to report this error...
+        timer.schedule_with_delay(duration.into(), move || {
+            let _ = sender.send(());
+        })
+    }
+}
+
+/// Convert a ScriptManagerError into an API Error.
+/// We can't implement From<T> because ScriptManagerError is in a different crate.
+fn sm_error(e: ScriptManagerError) -> Error {
+    Error::InternalError(InternalError::GenericError(format!("{:?}", e)))
+}
+
+impl<T> Adapter for ThinkerbellAdapter<T> where T: AdapterManagerHandle + API + Clone + Send + 'static {
+    fn id(&self) -> Id<AdapterId> {
+        self.adapter_id.clone()
+    }
+
+    fn name(&self) -> &str {
+        ADAPTER_NAME
+    }
+
+    fn vendor(&self) -> &str {
+        ADAPTER_VENDOR
+    }
+
+    fn version(&self) -> &[u32;4] {
+        &ADAPTER_VERSION
+    }
+
+    fn fetch_values(&self, set: Vec<Id<Getter>>) -> ResultMap<Id<Getter>, Option<Value>, Error> {
+        set.iter().map(|id| {
+            let (tx, rx) = channel();
```

Comment:
For better parallelization, I'd suggest two steps:

1. first, for all items of `set`, create `(tx, rx) and make the `RespondToGetter` calls, putting the `rx` in an array;
2. then, once step 1. is complete, listen to all the elements of the array.

Also using `drain(..)` instead of `iter()` will let you save one `id.clone()`.

All of this can be done in a followup, since we're in a hurry. My bad (and github's) for not seeing your PR.